### PR TITLE
correlation and covariance

### PR DESCRIPTION
**Pull request by jan-www**
This pull request implement the calculation of correlation and covariance bewteen two 1-d arrays.

Covariance measures the direction of a relationship between two variables, while correlation measures the strength of that relationship. Both correlation and covariance are positive when the variables move in the same direction, and negative when they move in opposite directions. However, a correlation coefficient must always be between -1 and +1, with the extreme values indicating a strong relationship.

### covariance
For a set of n data points with two variables x and y, the covariance is measured by taking the difference between each x and y variable and their respective means. These differences are then multiplied together, and averaged across all of the data points. In mathematical notation, this is expressed as:
image

### correlation
In [statistics](https://en.wikipedia.org/wiki/Statistics), correlation or dependence is any statistical relationship, whether [causal](https://en.wikipedia.org/wiki/Causality) or not, between two [random variables](https://en.wikipedia.org/wiki/Random_variable) or [bivariate data](https://en.wikipedia.org/wiki/Bivariate_data).

image

### CODE REVIEW

File: src/stat/corcov.ts [hunk 1]
```diff
@@ -30,3 +30,45 @@ export const getCorrelationMatrix = (xData: Tensor | RecursiveArray<number>): Te
...
+  if (xTensor.shape[0] !== yTensor.shape[0]) {
+    throw new TypeError('Length of input array x and y should be the same');
+  }
+  if (xTensor.shape[0] === 1) {
+    throw new TypeError('Length of input array should be greater than 1');
+  }
+  const nData = xTensor.shape[0];
+  const xMean = mean(x);
+  const yMean = mean(y);
+  return divNoNan(sum(mul(sub(xTensor, xMean), sub(yTensor, yMean))), nData - 1).dataSync()[0];
```
Comment (rickycao-qy) [inline src/stat/corcov.ts:55]: there are some Intermediate variables, can be wrapped by tidy

File: src/stat/corcov.ts [hunk 2]
```diff
@@ -30,3 +30,45 @@ export const getCorrelationMatrix = (xData: Tensor | RecursiveArray<number>): Te
...
+ * @param y input array y of shape (n,)
+ * @returns correltaion of x and y
+ */
+export const getCorrelation = (x: Tensor1D | number[], y: Tensor1D | number[]): number => {
+  const xTensor = checkArray(x, 'float32', 1) as Tensor1D;
+  const yTensor = checkArray(y, 'float32', 1) as Tensor1D;
+  const covariance = getCovariance(xTensor, yTensor);
+  const xVariance = getVariance(x);
+  const yVariance = getVariance(y);
+  return divNoNan(covariance, sqrt(mul(xVariance, yVariance))).dataSync()[0];
```
Comment (rickycao-qy) [inline src/stat/corcov.ts:73]: ditto

**Review by rickycao-qy — CHANGES_REQUESTED** [review]

File: src/stat/data.ts [hunk 3]
```diff
@@ -27,12 +33,19 @@ export const getMean = (xData: Tensor | RecursiveArray<number>): Tensor => {
...
+    const xTensor = checkArray(xData, 'float32');
+    if (xTensor.shape.length === 1) {
+      const xCentered = sub(xTensor, mean(xTensor));
+      const xStd = sqrt(getVariance(xTensor));
+      return divNoNan(xCentered, xStd);
+    } else {
+      const xCentered = sub(xTensor, mean(xTensor, axis));
+      const xStd = sqrt(getVariance(xTensor));
+      return divNoNan(xCentered, xStd);
+    }
```
Comment (FeelyChau) [inline src/stat/data.ts:47]:
```suggestion
   const xCentered = xTensor.shape.length === 1 ? sub(xTensor, mean(xTensor) : sub(xTensor, mean(xTensor, axis);
   const xStd = sqrt(getVariance(xTensor));
   return divNoNan(xCentered, xStd);
```

**Review by FeelyChau — CHANGES_REQUESTED** [review]

**Review by rickycao-qy — APPROVED** [review]
LGTM

**Pull request merged**